<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Import — import • pipette</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Import — import" />
<meta property="og:description" content="Read file by extension into R." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pipette</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.4.23</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://acidgenomics.com/">
    <span class="fas fa-vial fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/acidgenomics/r-pipette/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Import</h1>
    <small class="dont-index">Source: <a href='https://github.com/acidgenomics/r-pipette/blob/master/R/import.R'><code>R/import.R</code></a></small>
    <div class="hidden name"><code>import.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Read file by extension into R.</p>
    </div>

    <pre class="usage"><span class='fu'>import</span><span class='op'>(</span>
  <span class='va'>file</span>,
  format <span class='op'>=</span> <span class='st'>"auto"</span>,
  rownames <span class='op'>=</span> <span class='cn'>TRUE</span>,
  colnames <span class='op'>=</span> <span class='cn'>TRUE</span>,
  sheet <span class='op'>=</span> <span class='fl'>1L</span>,
  skip <span class='op'>=</span> <span class='fl'>0L</span>,
  makeNames <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/options.html'>getOption</a></span><span class='op'>(</span><span class='st'>"acid.import.make.names"</span>, default <span class='op'>=</span> <span class='fu'>syntactic</span><span class='fu'>::</span><span class='va'><a href='https://syntactic.acidgenomics.com/reference/makeNames.html'>makeNames</a></span><span class='op'>)</span>,
  metadata <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/options.html'>getOption</a></span><span class='op'>(</span><span class='st'>"acid.import.metadata"</span>, default <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>,
  quiet <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/options.html'>getOption</a></span><span class='op'>(</span><span class='st'>"acid.quiet"</span>, default <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>file</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
File path.</p></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
An optional file format type, which can be used to override the file format
inferred from <code>file</code>. Only recommended for file and URL paths that don't
contain an extension.</p></td>
    </tr>
    <tr>
      <th>rownames</th>
      <td><p><code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code>.
Automatically assign row names, if <code>rowname</code> column is defined.
Applies to file types that return <code>data.frame</code> only.</p></td>
    </tr>
    <tr>
      <th>colnames</th>
      <td><p><code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code> or <code>character</code>.
Automatically assign column names, using the first header row.
Applies to file types that return <code>data.frame</code> only.
Pass in a <code>character</code> vector to define the column names manually.</p></td>
    </tr>
    <tr>
      <th>sheet</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> or <code><a href='https://rdrr.io/r/base/integer.html'>integer(1)</a></code>.
<em>Applies to Excel Workbook, Google Sheet, or GraphPad Prism file.</em>
Sheet to read. Either a string (the name of a sheet), or an integer (the
position of the sheet). Defaults to the first sheet.</p></td>
    </tr>
    <tr>
      <th>skip</th>
      <td><p><code><a href='https://rdrr.io/r/base/integer.html'>integer(1)</a></code>.
<em>Applies to delimited file (CSV, TSV), Excel Workbook, or lines.</em>
Number of lines to skip.</p></td>
    </tr>
    <tr>
      <th>makeNames</th>
      <td><p><code>function</code>.
Apply syntactic naming function to (column) names.</p></td>
    </tr>
    <tr>
      <th>metadata</th>
      <td><p><code>list</code>.
Metadata.</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p><code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code>.
Perform command quietly, suppressing messages.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Varies, depending on the file type (format):</p><ul>
<li><p><strong>Plain text delimited</strong> (<code>CSV</code>, <code>TSV</code>, <code>TXT</code>): <code>data.frame</code>.<br />
Data separated by commas, tabs, or visual spaces.<br />
Note that TXT structure is amgibuous and actively discouraged.<br />
Refer to <code>Data frame return</code> section for details on how to change the
default return type to <code>DataFrame</code>, <code>tbl_df</code> or <code>data.table</code>.<br />
Imported by <code><a href='https://vroom.r-lib.org/reference/vroom.html'>vroom::vroom()</a></code>.</p></li>
<li><p><strong>Excel workbook</strong> (<code>XLSB</code>, <code>XLSX</code>): <code>data.frame</code>.<br />
Resave in plain text delimited format instead, if possible.<br />
Imported by <code><a href='https://readxl.tidyverse.org/reference/read_excel.html'>readxl::read_excel()</a></code>.</p></li>
<li><p><strong>Legacy Excel workbook (pre-2007)</strong> (<code>XLS</code>): <code>data.frame</code>.<br />
Resave in plain text delimited format instead, if possible.<br />
Note that import of files in this format is slow.<br />
Imported by <code><a href='https://readxl.tidyverse.org/reference/read_excel.html'>readxl::read_excel()</a></code>.</p></li>
<li><p><strong>GraphPad Prism project</strong> (<code>PZFX</code>): <code>data.frame</code>.<br />
Experimental. Consider resaving in CSV format instead.<br />
Imported by <code><a href='https://rdrr.io/pkg/pzfx/man/read_pzfx.html'>pzfx::read_pzfx()</a></code>.</p></li>
<li><p><strong>General feature format</strong> (<code>GFF</code>, <code>GFF1</code>, <code>GFF2</code>, <code>GFF3</code>, <code>GTF</code>):
<code>GRanges</code>.<br />
Imported by <code><a href='https://rdrr.io/pkg/rtracklayer/man/export.html'>rtracklayer::import()</a></code>.</p></li>
<li><p><strong>MatrixMarket exchange sparse matrix</strong> (<code>MTX</code>): <code>sparseMatrix</code>.<br />
Imported by <code><a href='https://rdrr.io/pkg/Matrix/man/externalFormats.html'>Matrix::readMM()</a></code>.</p></li>
<li><p><strong>Gene sets (for GSEA)</strong> (<code>GMT</code>, <code>GMX</code>): <code>character</code>.</p></li>
<li><p><strong>Browser extensible data</strong> (<code>BED</code>, <code>BED15</code>, <code>BEDGRAPH</code>, <code>BEDPE</code>):
<code>GRanges</code>.<br />
Imported by <code><a href='https://rdrr.io/pkg/rtracklayer/man/export.html'>rtracklayer::import()</a></code>.</p></li>
<li><p><strong>ChIP-seq peaks</strong> (<code>BROADPEAK</code>, <code>NARROWPEAK</code>): <code>GRanges</code>.<br />
Imported by <code><a href='https://rdrr.io/pkg/rtracklayer/man/export.html'>rtracklayer::import()</a></code>.</p></li>
<li><p><strong>Wiggle track format</strong> (<code>BIGWIG</code>, <code>BW</code>, <code>WIG</code>): <code>GRanges</code>.<br />
Imported by <code><a href='https://rdrr.io/pkg/rtracklayer/man/export.html'>rtracklayer::import()</a></code>.</p></li>
<li><p><strong>JSON serialization data</strong> (<code>JSON</code>): <code>list</code>.<br />
Imported by <code><a href='https://rdrr.io/pkg/jsonlite/man/read_json.html'>jsonlite::read_json()</a></code>.</p></li>
<li><p><strong>YAML serialization data</strong> (<code>YAML</code>, <code>YML</code>): <code>list</code>.<br />
Imported by <code><a href='https://rdrr.io/pkg/yaml/man/yaml.load.html'>yaml::yaml.load_file()</a></code>.</p></li>
<li><p><strong>Lines</strong> (<code>LOG</code>, <code>MD</code>, <code>PY</code>, <code>R</code>, <code>RMD</code>, <code>SH</code>): <code>character</code>.
Source code or log files.<br />
Imported by <code><a href='https://vroom.r-lib.org/reference/vroom_lines.html'>vroom_lines()</a></code>.</p></li>
<li><p><strong>R data serialized</strong> (<code>RDS</code>): <em>variable</em>.<br />
Currently recommend over RDA, if possible.<br />
Imported by <code><a href='https://rdrr.io/r/base/readRDS.html'>readRDS()</a></code>.</p></li>
<li><p><strong>R data</strong> (<code>RDA</code>, <code>RDATA</code>): <em>variable</em>.<br />
Must contain a single object.
Doesn't require internal object name to match, unlike <code><a href='loadData.html'>loadData()</a></code>.<br />
Imported by <code><a href='https://rdrr.io/r/base/load.html'>load()</a></code>.</p></li>
<li><p><strong>Infrequently used rio-compatible formats</strong> (<code>ARFF</code>, <code>DBF</code>, <code>DIF</code>, <code>DTA</code>,
<code>MAT</code>, <code>MTP</code>, <code>ODS</code>, <code>POR</code>, <code>SAS7BDAT</code>, <code>SAV</code>, <code>SYD</code>, <code>REC</code>, <code>XPT</code>):
<em>variable</em>.<br />
Imported by <code><a href='https://rdrr.io/pkg/rio/man/import.html'>rio::import()</a></code>.</p></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>import()</code> supports automatic loading of common file types, by wrapping
popular importer functions. It intentionally designed to be simple, with few
arguments. Remote URLs and compressed files are supported. If you need more
complex import settings, just call the wrapped importer directly instead.</p>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Updated 2021-01-05.</p>
    <h2 class="hasAnchor" id="row-and-column-names"><a class="anchor" href="#row-and-column-names"></a>Row and column names</h2>

    


<p><strong>Row names.</strong> Row name handling has become an inconsistent mess in R because
of differential support in base R, tidyverse, data.table, and Bioconductor.
To maintain sanity, <code>import()</code> attempts to handle row names automatically.
The function checks for a <code>rowname</code> column in delimited data, and moves these
values into the object's row names, if supported by the return type (e.g.
<code>data.frame</code>, <code>DataFrame</code>). Note that <code>tbl_df</code> (tibble) and <code>data.table</code>
intentionally do not support row names. When returning in this format, no
attempt to assign the <code>rowname</code> column into the return object's row names is
made. Note that <code>import()</code> is strict about this matching and only checks for
a <code>rowname</code> column, similar to the default syntax recommended in
<code><a href='https://tibble.tidyverse.org/reference/rownames.html'>tibble::rownames_to_column()</a></code>. To disable this behavior, set
<code>rownames = FALSE</code>, and no attempt will be made to set the row names.</p>
<p><strong>Column names.</strong> <code>import()</code> assumes that delimited files always contain
column names. If you are working with a file that doesn't contain column
names, either set <code>colnames = FALSE</code> or pass the names in as a <code>character</code>
vector. It's strongly recommended to always define column names in a
supported file type.</p>
    <h2 class="hasAnchor" id="data-frame-return"><a class="anchor" href="#data-frame-return"></a>Data frame return</h2>

    


<p>By default, <code>import()</code> returns a standard <code>data.frame</code> for delimited/column
formatted data. However, any of these desired output formats can be set
globally using <code><a href='https://rdrr.io/r/base/options.html'>options(acid.data.frame = "data.frame")</a></code>.</p>
<p>Supported return types:</p><ul>
<li><p><code>data.frame</code>: Base R default. Generally recommended.</p><ul>
<li><p>S3 class.</p></li>
<li><p>Allows rownames, but they're required and can't be set <code>NULL</code>.</p></li>
<li><p>See <code><a href='https://rdrr.io/r/base/data.frame.html'>help(topic = "data.frame", package = "base")</a></code> for details.</p></li>
</ul></li>
<li><p><code>DataFrame</code>: Recommended when working with Bioconductor packages.</p><ul>
<li><p>S4 class.</p></li>
<li><p>Allows rownames, but they're optional and can be set <code>NULL</code>.</p></li>
<li><p>See <code><a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>help(topic = "DataFrame", package = "S4Vectors")</a></code> for details.</p></li>
</ul></li>
<li><p><code>tbl_df</code> (<code>tibble</code>): Recommended when working with tidyverse packages.</p><ul>
<li><p>S3 class; inherits <code>data.frame</code>.</p></li>
<li><p>Does not allow rownames.</p></li>
<li><p>See <code><a href='https://tibble.tidyverse.org/reference/tibble.html'>help(topic = "tibble", package = "tibble")</a></code> for details.</p></li>
</ul></li>
<li><p><code>data.table</code>: Recommended when working with the data.table package.</p><ul>
<li><p>S3 class; inherits <code>data.frame</code>.</p></li>
<li><p>Does not allow rownames.</p></li>
<li><p>See <code><a href='https://Rdatatable.gitlab.io/data.table/reference/data.table.html'>help(topic = "data.table", package = "data.table")</a></code> for details.</p></li>
</ul></li>
</ul>

<p>Note that <code>stringsAsFactors</code> is always disabled for import.</p>
    <h2 class="hasAnchor" id="matrix-market-exchange-mtx-"><a class="anchor" href="#matrix-market-exchange-mtx-"></a>Matrix Market Exchange (MTX)</h2>

    


<p>Reading a Matrix Market Exchange file requires <code>ROWNAMES</code> and <code>COLNAMES</code>
sidecar files containing the corresponding row and column names of the sparse
matrix.</p>
    <h2 class="hasAnchor" id="general-feature-format-gff-gtf-"><a class="anchor" href="#general-feature-format-gff-gtf-"></a>General feature format (GFF, GTF)</h2>

    


<p>The GFF (General Feature Format) format consists of one line per feature,
each containing 9 columns of data, plus optional track definition lines. The
GTF (General Transfer Format) is identical to GFF version 2.</p>
<p><a href='https://basejump.acidgenomics.com/'>basejump</a> exports the specialized <code>makeGRangesFromGFF()</code> function that
makes GFF loading simple.</p>
<p>See also:</p><ul>
<li><p><a href='http://www.ensembl.org/info/website/upload/gff.html'>Ensembl spec</a></p></li>
<li><p><a href='http://www.gencodegenes.org/gencodeformat.html'>GENCODE spec</a></p></li>
</ul>

    <h2 class="hasAnchor" id="gene-sets-gmt-gmx-"><a class="anchor" href="#gene-sets-gmt-gmx-"></a>Gene sets (GMT, GMX)</h2>

    


<p>Refer to the Broad Institute <a href='https://goo.gl/3ZkDPb'>GSEA wiki</a> for details.</p>
    <h2 class="hasAnchor" id="bcbio-count-matrix"><a class="anchor" href="#bcbio-count-matrix"></a>bcbio count matrix</h2>

    


<p><a href='https://bcbio-nextgen.readthedocs.io/'>bcbio</a> count matrix and related sidecar files are natively supported.</p><ul>
<li><p><code>COUNTS</code>: Counts table (e.g. RNA-seq aligned counts).</p></li>
<li><p><code>COLNAMES</code>: Sidecar file containing column names.</p></li>
<li><p><code>ROWNAMES</code>: Sidecar file containing row names.</p></li>
</ul>

    <h2 class="hasAnchor" id="blacklisted-extensions"><a class="anchor" href="#blacklisted-extensions"></a>Blacklisted extensions</h2>

    


<p>These file formats are blacklisted, and intentionally not supported:
<code>DOC</code>, <code>DOCX</code>, <code>PDF</code>, <code>PPT</code>, <code>PPTX</code>.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Packages:</p><ul>
<li><p><a href='http://r-datatable.com/'>data.table</a>.</p></li>
<li><p><a href='http://readr.tidyverse.org'>readr</a>.</p></li>
<li><p><a href='http://readxl.tidyverse.org'>readxl</a>.</p></li>
<li><p><a href='https://cran.r-project.org/package=rio'>rio</a>.</p></li>
<li><p><a href='http://bioconductor.org/packages/rtracklayer/'>rtracklayer</a>.</p></li>
<li><p><a href='https://vroom.r-lib.org'>vroom</a>.</p></li>
</ul>

<p>Import functions:</p><ul>
<li><p><code><a href='https://Rdatatable.gitlab.io/data.table/reference/fread.html'>data.table::fread()</a></code>.</p></li>
<li><p><code><a href='https://readr.tidyverse.org/reference/read_delim.html'>readr::read_csv()</a></code>.</p></li>
<li><p><code><a href='https://rdrr.io/pkg/rio/man/import.html'>rio::import()</a></code>.</p></li>
<li><p><code><a href='https://rdrr.io/pkg/rtracklayer/man/export.html'>rtracklayer::import()</a></code>.</p></li>
<li><p><code><a href='https://rdrr.io/r/utils/read.table.html'>utils::read.table()</a></code>.</p></li>
<li><p><code><a href='https://vroom.r-lib.org/reference/vroom.html'>vroom::vroom()</a></code>.</p></li>
</ul>
</div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='va'>file</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/example.csv"</span>, package <span class='op'>=</span> <span class='st'>"pipette"</span><span class='op'>)</span>

<span class='co'>## Row and column names enabled.</span>
<span class='va'>x</span> <span class='op'>&lt;-</span> <span class='fu'>import</span><span class='op'>(</span><span class='va'>file</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>→ Importing <span style='color: #0000BB;'>example.csv</span><span> at </span><span style='color: #0000BB;'>/private/var/folders/l1/8y8sjzmn15v49jgrqglghcfr0000gn/T/RtmpGRad5U/temp_libpath1076b1a12963e/pipette/extdata</span><span> using </span><span style='color: #0000BB;'>vroom</span><span>::</span><span style='color: #000000;background-color: #BBBBBB;'>`vroom()`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> Setting row names from </span><span style='color: #000000;background-color: #BBBBBB;'>`rowname`</span><span> column.</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt;       sample1 sample2 sample3 sample4
#&gt; gene1      16      20      13      16
#&gt; gene2      29      22      43      50
#&gt; gene3     243     245     186     184
#&gt; gene4       7      14      25      16
#&gt; gene5       1       1       2       2
#&gt; gene6      10      17      18      11</div><div class='input'>
<span class='co'>## Row and column names disabled.</span>
<span class='va'>x</span> <span class='op'>&lt;-</span> <span class='fu'>import</span><span class='op'>(</span><span class='va'>file</span>, rownames <span class='op'>=</span> <span class='cn'>FALSE</span>, colnames <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>→ Importing </span><span style='color: #0000BB;'>example.csv</span><span> at </span><span style='color: #0000BB;'>/private/var/folders/l1/8y8sjzmn15v49jgrqglghcfr0000gn/T/RtmpGRad5U/temp_libpath1076b1a12963e/pipette/extdata</span><span> using </span><span style='color: #0000BB;'>vroom</span><span>::</span><span style='color: #000000;background-color: #BBBBBB;'>`vroom()`</span><span>.</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt;        X1      X2      X3      X4      X5
#&gt; 1 rowname sample1 sample2 sample3 sample4
#&gt; 2   gene1      16      20      13      16
#&gt; 3   gene2      29      22      43      50
#&gt; 4   gene3     243     245     186     184
#&gt; 5   gene4       7      14      25      16
#&gt; 6   gene5       1       1       2       2</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


